File: frontend/src/app/core/services/forum.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, catchError, throwError } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  ForumThread,
  ForumPost,
  CreateThreadRequest,
  CreatePostRequest,
  PaginatedResponse,
} from '../../models/domain.model';

export interface ForumFilterOptions {
  category?: string;
  page?: number;
  size?: number;
  sort?: string;
}

@Injectable({
  providedIn: 'root',
})
export class ForumService {
  private baseUrl = `${environment.apiUrl}/forum`;

  constructor(private http: HttpClient) {}

  // ========== Thread Operations ==========

  /**
   * Get all forum threads with pagination
   */
  getAllThreads(filter?: ForumFilterOptions): Observable<PaginatedResponse<ForumThread>> {
    let params = new HttpParams();

    if (filter) {
      if (filter.category) params = params.set('category', filter.category);
      if (filter.page !== undefined) params = params.set('page', filter.page.toString());
      if (filter.size !== undefined) params = params.set('size', filter.size.toString());
      if (filter.sort) params = params.set('sort', filter.sort);
    } else {
      params = params.set('page', '0').set('size', '10');
    }

    return this.http
      .get<PaginatedResponse<ForumThread>>(`${this.baseUrl}/threads`, { params })
      .pipe(
        catchError((error) => {
          console.error('Error fetching threads:', error);
          return throwError(() => ({
            message: error.error?.message || 'Failed to fetch threads',
            status: error.status,
          }));
        })
      );
  }

  /**
   * Get threads by category
   */
  getThreadsByCategory(
    category: string,
    page = 0,
    size = 10
  ): Observable<PaginatedResponse<ForumThread>> {
    const params = new HttpParams()
      .set('category', category)
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http
      .get<PaginatedResponse<ForumThread>>(`${this.baseUrl}/threads`, { params })
      .pipe(
        catchError((error) => {
          console.error('Error fetching threads by category:', error);
          return throwError(() => ({
            message: error.error?.message || 'Failed to fetch threads',
            status: error.status,
          }));
        })
      );
  }

  /**
   * Get thread by ID
   */
  getThread(id: string): Observable<ForumThread> {
    return this.http.get<ForumThread>(`${this.baseUrl}/threads/${id}`).pipe(
      catchError((error) => {
        console.error('Error fetching thread:', error);
        return throwError(() => ({
          message: error.error?.message || 'Failed to fetch thread',
          status: error.status,
        }));
      })
    );
  }

  /**
   * Create new thread
   */
  createThread(data: CreateThreadRequest): Observable<ForumThread> {
    return this.http.post<ForumThread>(`${this.baseUrl}/threads`, data).pipe(
      catchError((error) => {
        console.error('Error creating thread:', error);
        return throwError(() => ({
          message: error.error?.message || 'Failed to create thread',
          status: error.status,
        }));
      })
    );
  }

  /**
   * Update thread
   */
  updateThread(id: string, data: Partial<ForumThread>): Observable<ForumThread> {
    return this.http.put<ForumThread>(`${this.baseUrl}/threads/${id}`, data).pipe(
      catchError((error) => {
        console.error('Error updating thread:', error);
        return throwError(() => ({
          message: error.error?.message || 'Failed to update thread',
          status: error.status,
        }));
      })
    );
  }

  /**
   * Delete thread
   */
  deleteThread(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/threads/${id}`).pipe(
      catchError((error) => {
        console.error('Error deleting thread:', error);
        return throwError(() => ({
          message: error.error?.message || 'Failed to delete thread',
          status: error.status,
        }));
      })
    );
  }

  /**
   * Pin/unpin thread
   */
  togglePin(id: string): Observable<ForumThread> {
    return this.http.post<ForumThread>(`${this.baseUrl}/threads/${id}/pin`, {}).pipe(
      catchError((error) => {
        console.error('Error toggling pin:', error);
        return throwError(() => ({
          message: error.error?.message || 'Failed to toggle pin',
          status: error.status,
        }));
      })
    );
  }

  /**
   * Lock/unlock thread
   */
  toggleLock(id: string): Observable<ForumThread> {
    return this.http.post<ForumThread>(`${this.baseUrl}/threads/${id}/lock`, {}).pipe(
      catchError((error) => {
        console.error('Error toggling lock:', error);
        return throwError(() => ({
          message: error.error?.message || 'Failed to toggle lock',
          status: error.status,
        }));
      })
    );
  }

  // ========== Post Operations ==========

  /**
   * Get posts in thread
   */
  getThreadPosts(threadId: string, page = 0, size = 10): Observable<PaginatedResponse<ForumPost>> {
    const params = new HttpParams().set('page', page.toString()).set('size', size.toString());
    return this.http
      .get<PaginatedResponse<ForumPost>>(`${this.baseUrl}/threads/${threadId}/posts`, { params })
      .pipe(
        catchError((error) => {
          console.error('Error fetching thread posts:', error);
          return throwError(() => ({
            message: error.error?.message || 'Failed to fetch posts',
            status: error.status,
          }));
        })
      );
  }

  /**
   * Get post by ID
   */
  getPost(postId: string): Observable<ForumPost> {
    return this.http.get<ForumPost>(`${this.baseUrl}/posts/${postId}`).pipe(
      catchError((error) => {
        console.error('Error fetching post:', error);
        return throwError(() => ({
          message: error.error?.message || 'Failed to fetch post',
          status: error.status,
        }));
      })
    );
  }

  /**
   * Create new post in thread
   */
  createPost(data: CreatePostRequest): Observable<ForumPost> {
    return this.http.post<ForumPost>(`${this.baseUrl}/posts`, data).pipe(
      catchError((error) => {
        console.error('Error creating post:', error);
        return throwError(() => ({
          message: error.error?.message || 'Failed to create post',
          status: error.status,
        }));
      })
    );
  }

  /**
   * Update post
   */
  updatePost(id: string, content: string): Observable<ForumPost> {
    return this.http.put<ForumPost>(`${this.baseUrl}/posts/${id}`, { content }).pipe(
      catchError((error) => {
        console.error('Error updating post:', error);
        return throwError(() => ({
          message: error.error?.message || 'Failed to update post',
          status: error.status,
        }));
      })
    );
  }

  /**
   * Delete post
   */
  deletePost(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/posts/${id}`).pipe(
      catchError((error) => {
        console.error('Error deleting post:', error);
        return throwError(() => ({
          message: error.error?.message || 'Failed to delete post',
          status: error.status,
        }));
      })
    );
  }

  // ========== User Operations ==========

  /**
   * Get user's threads
   */
  getUserThreads(page = 0, size = 10): Observable<PaginatedResponse<ForumThread>> {
    const params = new HttpParams().set('page', page.toString()).set('size', size.toString());
    return this.http
      .get<PaginatedResponse<ForumThread>>(`${this.baseUrl}/user/threads`, { params })
      .pipe(
        catchError((error) => {
          console.error('Error fetching user threads:', error);
          return throwError(() => ({
            message: error.error?.message || 'Failed to fetch user threads',
            status: error.status,
          }));
        })
      );
  }

  /**
   * Get user's posts
   */
  getUserPosts(page = 0, size = 10): Observable<PaginatedResponse<ForumPost>> {
    const params = new HttpParams().set('page', page.toString()).set('size', size.toString());
    return this.http
      .get<PaginatedResponse<ForumPost>>(`${this.baseUrl}/user/posts`, { params })
      .pipe(
        catchError((error) => {
          console.error('Error fetching user posts:', error);
          return throwError(() => ({
            message: error.error?.message || 'Failed to fetch user posts',
            status: error.status,
          }));
        })
      );
  }
}

File: backend/src/main/java/com/archipellibre/service/ForumService.java
package com.archipellibre.service;

import com.archipellibre.dto.ForumPostCreateRequest;
import com.archipellibre.dto.ForumPostUpdateRequest;
import com.archipellibre.dto.ForumThreadCreateRequest;
import com.archipellibre.dto.ForumThreadUpdateRequest;
import com.archipellibre.exception.BusinessLogicException;
import com.archipellibre.exception.ResourceNotFoundException;
import com.archipellibre.model.ForumPost;
import com.archipellibre.model.ForumThread;
import com.archipellibre.model.User;
import com.archipellibre.repository.ForumPostRepository;
import com.archipellibre.repository.ForumThreadRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
@Transactional
@Slf4j
@RequiredArgsConstructor
@SuppressWarnings("null")
public class ForumService {

    private final ForumThreadRepository threadRepository;
    private final ForumPostRepository postRepository;
    private final UserService userService;

    // ===== THREAD MANAGEMENT =====

    /**
     * Create a new forum thread
     */
    public ForumThread createThread(UUID creatorId, ForumThreadCreateRequest request) {
        User creator = userService.getUserById(creatorId);

        ForumThread thread = ForumThread.builder()
                .title(request.getTitle())
                .content(request.getContent())
                .category(request.getCategory())
                .creator(creator)
                .pinned(false)
                .locked(false)
                .lastActivityAt(LocalDateTime.now())
                .build();

        ForumThread savedThread = threadRepository.save(thread);
        log.info("Forum thread created: {} by user: {}", savedThread.getId(), creatorId);
        return savedThread;
    }

    /**
     * Get thread by ID
     */
    public ForumThread getThreadById(UUID threadId) {
        return threadRepository.findById(threadId)
                .orElseThrow(() -> new ResourceNotFoundException("Forum thread not found with ID: " + threadId));
    }

    /**
     * Get all threads with pagination (ordered by recent activity)
     */
    @Transactional(readOnly = true)
    public Page<ForumThread> getAllThreads(Pageable pageable) {
        return threadRepository.findAllByOrderByLastActivityAtDesc(pageable);
    }

    /**
     * Get threads by category
     */
    @Transactional(readOnly = true)
    public Page<ForumThread> getThreadsByCategory(String category, Pageable pageable) {
        return threadRepository.findByCategory(category, pageable);
    }

    /**
     * Get threads created by user
     */
    @Transactional(readOnly = true)
    public Page<ForumThread> getThreadsByCreator(UUID creatorId, Pageable pageable) {
        userService.getUserById(creatorId); // Verify user exists
        return threadRepository.findByCreatorId(creatorId, pageable);
    }

    /**
     * Get pinned threads (sticky)
     */
    @Transactional(readOnly = true)
    public List<ForumThread> getPinnedThreads() {
        return threadRepository.findByPinnedTrue();
    }

    /**
     * Update thread details (creator or moderator only)
     */
    public ForumThread updateThread(UUID threadId, ForumThreadUpdateRequest request) {
        ForumThread thread = getThreadById(threadId);

        if (request.getTitle() != null) {
            thread.setTitle(request.getTitle());
        }

        if (request.getContent() != null) {
            thread.setContent(request.getContent());
        }

        if (request.getCategory() != null) {
            thread.setCategory(request.getCategory());
        }

        ForumThread updated = threadRepository.save(thread);
        log.info("Forum thread updated: {}", threadId);
        return updated;
    }

    /**
     * Pin thread (sticky thread - moderator/admin only)
     */
    public ForumThread pinThread(UUID threadId) {
        ForumThread thread = getThreadById(threadId);

        if (thread.getPinned()) {
            throw new BusinessLogicException("Thread is already pinned");
        }

        thread.setPinned(true);
        ForumThread updated = threadRepository.save(thread);
        log.info("Thread pinned: {}", threadId);
        return updated;
    }

    /**
     * Unpin thread
     */
    public ForumThread unpinThread(UUID threadId) {
        ForumThread thread = getThreadById(threadId);

        if (!thread.getPinned()) {
            throw new BusinessLogicException("Thread is not pinned");
        }

        thread.setPinned(false);
        ForumThread updated = threadRepository.save(thread);
        log.info("Thread unpinned: {}", threadId);
        return updated;
    }

    /**
     * Lock thread (no new posts allowed - moderator/admin only)
     */
    public ForumThread lockThread(UUID threadId, String reason) {
        ForumThread thread = getThreadById(threadId);

        if (thread.getLocked()) {
            throw new BusinessLogicException("Thread is already locked");
        }

        thread.setLocked(true);
        ForumThread updated = threadRepository.save(thread);
        log.info("Thread locked: {} - Reason: {}", threadId, reason);
        return updated;
    }

    /**
     * Unlock thread
     */
    public ForumThread unlockThread(UUID threadId) {
        ForumThread thread = getThreadById(threadId);

        if (!thread.getLocked()) {
            throw new BusinessLogicException("Thread is not locked");
        }

        thread.setLocked(false);
        ForumThread updated = threadRepository.save(thread);
        log.info("Thread unlocked: {}", threadId);
        return updated;
    }

    /**
     * Check if thread is locked
     */
    @Transactional(readOnly = true)
    public boolean isThreadLocked(UUID threadId) {
        ForumThread thread = getThreadById(threadId);
        return thread.getLocked();
    }

    /**
     * Delete thread (moderator/admin only)
     */
    public void deleteThread(UUID threadId) {
        ForumThread thread = getThreadById(threadId);
        threadRepository.delete(thread);
        log.info("Forum thread deleted: {}", threadId);
    }

    /**
     * Get thread statistics
     */
    @Transactional(readOnly = true)
    public ThreadStats getThreadStats(UUID threadId) {
        ForumThread thread = getThreadById(threadId);
        long postCount = postRepository.countByThreadId(threadId);

        return ThreadStats.builder()
                .threadId(threadId)
                .title(thread.getTitle())
                .category(thread.getCategory())
                .postCount(postCount)
                .pinned(thread.getPinned())
                .locked(thread.getLocked())
                .createdBy(thread.getCreator().getUsername())
                .createdAt(thread.getCreatedAt())
                .lastActivityAt(thread.getLastActivityAt())
                .build();
    }

    // ===== POST MANAGEMENT =====

    /**
     * Create a post in a thread
     */
    public ForumPost createPost(UUID threadId, UUID authorId, ForumPostCreateRequest request) {
        ForumThread thread = getThreadById(threadId);
        User author = userService.getUserById(authorId);

        if (thread.getLocked()) {
            throw new BusinessLogicException("Cannot post in a locked thread");
        }

        ForumPost post = ForumPost.builder()
                .content(request.getContent())
                .author(author)
                .thread(thread)
                .edited(false)
                .build();

        ForumPost savedPost = postRepository.save(post);
        
        // Update thread's last activity
        thread.setLastActivityAt(LocalDateTime.now());
        threadRepository.save(thread);

        log.info("Forum post created: {} in thread: {} by user: {}", savedPost.getId(), threadId, authorId);
        return savedPost;
    }

    /**
     * Get post by ID
     */
    public ForumPost getPostById(UUID postId) {
        return postRepository.findById(postId)
                .orElseThrow(() -> new ResourceNotFoundException("Forum post not found with ID: " + postId));
    }

    /**
     * Get posts in thread
     */
    @Transactional(readOnly = true)
    public Page<ForumPost> getPostsByThread(UUID threadId, Pageable pageable) {
        getThreadById(threadId); // Verify thread exists
        return postRepository.findByThreadId(threadId, pageable);
    }

    /**
     * Get posts by author
     */
    @Transactional(readOnly = true)
    public Page<ForumPost> getPostsByAuthor(UUID authorId, Pageable pageable) {
        userService.getUserById(authorId); // Verify user exists
        return postRepository.findByAuthorId(authorId, pageable);
    }

    /**
     * Update post (author or moderator only)
     */
    public ForumPost updatePost(UUID postId, ForumPostUpdateRequest request) {
        ForumPost post = getPostById(postId);

        if (request.getContent() != null) {
            post.setContent(request.getContent());
            post.setEdited(true);
        }

        ForumPost updated = postRepository.save(post);
        log.info("Forum post updated: {}", postId);
        return updated;
    }

    /**
     * Delete post (author or moderator only)
     */
    public void deletePost(UUID postId) {
        ForumPost post = getPostById(postId);
        ForumThread thread = post.getThread();
        
        postRepository.delete(post);
        
        // Update thread's last activity
        thread.setLastActivityAt(LocalDateTime.now());
        threadRepository.save(thread);

        log.info("Forum post deleted: {}", postId);
    }

    /**
     * Get post count in thread
     */
    @Transactional(readOnly = true)
    public long getPostCount(UUID threadId) {
        getThreadById(threadId); // Verify thread exists
        return postRepository.countByThreadId(threadId);
    }

    /**
     * Search threads by keyword (simple - can be enhanced with Elasticsearch later)
     */
    @Transactional(readOnly = true)
    public List<ForumThread> searchThreads(String keyword) {
        return threadRepository.findAll().stream()
                .filter(t -> 
                    t.getTitle().toLowerCase().contains(keyword.toLowerCase()) ||
                    t.getContent().toLowerCase().contains(keyword.toLowerCase())
                )
                .toList();
    }

    // ===== MODERATION ENDPOINTS =====

    /**
     * Flag post for moderation (report inappropriate content)
     */
    public void flagPostForModeration(UUID postId, String reason) {
        getPostById(postId); // Verify post exists
        // In a real application, this would create a moderation ticket
        log.warn("Post flagged for moderation: {} - Reason: {}", postId, reason);
    }

    /**
     * Get post statistics
     */
    @Transactional(readOnly = true)
    public PostStats getPostStats(UUID postId) {
        ForumPost post = getPostById(postId);

        return PostStats.builder()
                .postId(postId)
                .authorUsername(post.getAuthor().getUsername())
                .threadId(post.getThread().getId())
                .threadTitle(post.getThread().getTitle())
                .edited(post.getEdited())
                .createdAt(post.getCreatedAt())
                .updatedAt(post.getUpdatedAt())
                .contentLength(post.getContent().length())
                .build();
    }

    // ===== HELPER CLASSES =====

    @lombok.Data
    @lombok.Builder
    public static class ThreadStats {
        private UUID threadId;
        private String title;
        private String category;
        private long postCount;
        private boolean pinned;
        private boolean locked;
        private String createdBy;
        private LocalDateTime createdAt;
        private LocalDateTime lastActivityAt;
    }

    @lombok.Data
    @lombok.Builder
    public static class PostStats {
        private UUID postId;
        private String authorUsername;
        private UUID threadId;
        private String threadTitle;
        private boolean edited;
        private LocalDateTime createdAt;
        private LocalDateTime updatedAt;
        private int contentLength;
    }
}
